# Feature Engineering
## Data import

In [1]:
import numpy as np
import pandas as pd
offline = pd.read_csv('offline_train.csv', header=0)
print(offline.shape[0])

1754884


## Delete null value in the dataset

In [2]:
offline_notNull = offline[offline.Coupon_id.notnull()]
print(offline_notNull.shape[0])

1053282


In [3]:
offline_notNull

User_id  Merchant_id  Coupon_id Discount_rate  Distance  \
1        1439408         4663    11002.0        150:20       1.0   
2        1439408         2632     8591.0          20:1       0.0   
3        1439408         2632     1078.0          20:1       0.0   
4        1439408         2632     8591.0          20:1       0.0   
6        1439408         2632     8591.0          20:1       0.0   
...          ...          ...        ...           ...       ...   
1754877   212662         3021     3739.0          30:1       6.0   
1754878   212662         2934     5686.0          30:5       2.0   
1754880   212662         3021     3739.0          30:1       6.0   
1754882   752472         7113     1633.0         50:10       6.0   
1754883   752472         3621     2705.0          20:5       0.0   

         Date_received        Date  
1           20160528.0         NaN  
2           20160217.0         NaN  
3           20160319.0         NaN  
4           20160613.0         NaN  
6           20160516.0  20160613.0  
...                ...         ...  
1754877     20160504.0  20160508.0  
1754878     20160321.0  20160322.0  
1754880     20160508.0  20160602.0  
1754882     20160613.0         NaN  
1754883     20160523.0         NaN  

[1053282 rows x 7 columns]

These 1,053,282 is our predict datas
## The coupon classification

In [4]:
temp = offline_notNull
coupon_consume = temp[temp.Date.notnull()]
coupon_no_consume = temp[temp.Date.isnull()]
Ordinary_Consumption = offline[offline.Date.notnull() & offline.Date_received.isnull()]

The following sell provide the `coupon_consume` dataset preview (in DataShell)

In [5]:
coupon_consume

User_id  Merchant_id  Coupon_id Discount_rate  Distance  \
6        1439408         2632     8591.0          20:1       0.0   
33       1113008         1361    11166.0          20:1       0.0   
38       2881376         8390     7531.0          20:5       0.0   
69        114747         6901     2366.0          30:5       0.0   
75        114747         5341      111.0          30:5       0.0   
...          ...          ...        ...           ...       ...   
1754833  1437872         7706      416.0        100:10       4.0   
1754873   212662         2934     5686.0          30:5       2.0   
1754877   212662         3021     3739.0          30:1       6.0   
1754878   212662         2934     5686.0          30:5       2.0   
1754880   212662         3021     3739.0          30:1       6.0   

         Date_received        Date  
6           20160516.0  20160613.0  
33          20160515.0  20160521.0  
38          20160321.0  20160329.0  
69          20160523.0  20160605.0  
75          20160127.0  20160221.0  
...                ...         ...  
1754833     20160129.0  20160202.0  
1754873     20160321.0  20160330.0  
1754877     20160504.0  20160508.0  
1754878     20160321.0  20160322.0  
1754880     20160508.0  20160602.0  

[75382 rows x 7 columns]

The following sell provide the `conupon_no_consume` dataset preview (in Datashell)

In [6]:
coupon_no_consume

User_id  Merchant_id  Coupon_id Discount_rate  Distance  \
1        1439408         4663    11002.0        150:20       1.0   
2        1439408         2632     8591.0          20:1       0.0   
3        1439408         2632     1078.0          20:1       0.0   
4        1439408         2632     8591.0          20:1       0.0   
7        1832624         3381     7610.0        200:20       0.0   
...          ...          ...        ...           ...       ...   
1754866  3795296         5341      111.0          30:5       0.0   
1754869   188086         6568     4723.0          30:1       0.0   
1754876   212662         3532     5267.0          30:5       1.0   
1754882   752472         7113     1633.0         50:10       6.0   
1754883   752472         3621     2705.0          20:5       0.0   

         Date_received  Date  
1           20160528.0   NaN  
2           20160217.0   NaN  
3           20160319.0   NaN  
4           20160613.0   NaN  
7           20160429.0   NaN  
...                ...   ...  
1754866     20160129.0   NaN  
1754869     20160415.0   NaN  
1754876     20160322.0   NaN  
1754882     20160613.0   NaN  
1754883     20160523.0   NaN  

[977900 rows x 7 columns]

The following sell provide the `Ordinary_Consumption` dataset preview (in Datashell)

In [7]:
Ordinary_Consumption

User_id  Merchant_id  Coupon_id Discount_rate  Distance  \
0        1439408         2632        NaN           NaN       0.0   
5        1439408         2632        NaN           NaN       0.0   
11       2029232         6459        NaN           NaN       0.0   
12       2029232         6459        NaN           NaN       0.0   
14        196342         1579        NaN           NaN       1.0   
...          ...          ...        ...           ...       ...   
1754872   212662         3532        NaN           NaN       1.0   
1754874   212662         2934        NaN           NaN       2.0   
1754875   212662         2934        NaN           NaN       2.0   
1754879   212662         3532        NaN           NaN       1.0   
1754881   212662         2934        NaN           NaN       2.0   

         Date_received        Date  
0                  NaN  20160217.0  
5                  NaN  20160516.0  
11                 NaN  20160626.0  
12                 NaN  20160519.0  
14                 NaN  20160606.0  
...                ...         ...  
1754872            NaN  20160308.0  
1754874            NaN  20160513.0  
1754875            NaN  20160512.0  
1754879            NaN  20160322.0  
1754881            NaN  20160321.0  

[701602 rows x 7 columns]

## 1. Customer Feature selection

Customer is the main entity of the trade, therefore, some features about them is extremely important.

It seems that some the user have duplicated behaviors of a single customer.

Therefore, It is very important to learn about the usage problems.

In [8]:
# Frequency of use
temp = coupon_consume.groupby('User_id').size().reset_index(name='u1')
X = temp
X

User_id  u1
0          184   1
1          417   1
2          687   2
3          696   4
4          947   2
...        ...  ..
46390  7360797   1
46391  7360845   1
46392  7360863   1
46393  7360928   1
46394  7360941   2

[46395 rows x 2 columns]

Also, the frequency of not using is also worth being focused.

In [9]:
# frequency of not use
temp = coupon_no_consume.groupby('User_id').size().reset_index(name='u2')
X = pd.merge(X, temp, how='left',on='User_id')
X

User_id  u1   u2
0          184   1  1.0
1          417   1  4.0
2          687   2  6.0
3          696   4  5.0
4          947   2  4.0
...        ...  ..  ...
46390  7360797   1  2.0
46391  7360845   1  2.0
46392  7360863   1  2.0
46393  7360928   1  2.0
46394  7360941   2  NaN

[46395 rows x 3 columns]

Leave the NaN in the dataset, It will have the other usage

In [10]:
# What is the percentage that the user collect the coupon?
X['u3'] = X.u1.fillna(0) + X.u2.fillna(0)
X

User_id  u1   u2   u3
0          184   1  1.0  2.0
1          417   1  4.0  5.0
2          687   2  6.0  8.0
3          696   4  5.0  9.0
4          947   2  4.0  6.0
...        ...  ..  ...  ...
46390  7360797   1  2.0  3.0
46391  7360845   1  2.0  3.0
46392  7360863   1  2.0  3.0
46393  7360928   1  2.0  3.0
46394  7360941   2  NaN  2.0

[46395 rows x 4 columns]

In [11]:
# What is the percentage that the user use their coupon?
X['u4'] = X.u1 / X.u3
X

User_id  u1   u2   u3        u4
0          184   1  1.0  2.0  0.500000
1          417   1  4.0  5.0  0.200000
2          687   2  6.0  8.0  0.250000
3          696   4  5.0  9.0  0.444444
4          947   2  4.0  6.0  0.333333
...        ...  ..  ...  ...       ...
46390  7360797   1  2.0  3.0  0.333333
46391  7360845   1  2.0  3.0  0.333333
46392  7360863   1  2.0  3.0  0.333333
46393  7360928   1  2.0  3.0  0.333333
46394  7360941   2  NaN  2.0  1.000000

[46395 rows x 5 columns]

In [12]:
# The opposite: What is the percentage that the user won't use their coupon?
X ['u5'] = X.u2 / X.u3
X['u5'].fillna(0.0, inplace=True)
X

User_id  u1   u2   u3        u4        u5
0          184   1  1.0  2.0  0.500000  0.500000
1          417   1  4.0  5.0  0.200000  0.800000
2          687   2  6.0  8.0  0.250000  0.750000
3          696   4  5.0  9.0  0.444444  0.555556
4          947   2  4.0  6.0  0.333333  0.666667
...        ...  ..  ...  ...       ...       ...
46390  7360797   1  2.0  3.0  0.333333  0.666667
46391  7360845   1  2.0  3.0  0.333333  0.666667
46392  7360863   1  2.0  3.0  0.333333  0.666667
46393  7360928   1  2.0  3.0  0.333333  0.666667
46394  7360941   2  NaN  2.0  1.000000  0.000000

[46395 rows x 6 columns]

Interesting, these features is calculating the coupon's feature, However, how about their ordinary consumption??

In [13]:
# make a count
temp = Ordinary_Consumption.groupby('User_id').size().reset_index(name='u6')
X = pd.merge(X,temp,how='left',on='User_id')
X

User_id  u1   u2   u3        u4        u5    u6
0          184   1  1.0  2.0  0.500000  0.500000   NaN
1          417   1  4.0  5.0  0.200000  0.800000   2.0
2          687   2  6.0  8.0  0.250000  0.750000   NaN
3          696   4  5.0  9.0  0.444444  0.555556  26.0
4          947   2  4.0  6.0  0.333333  0.666667   5.0
...        ...  ..  ...  ...       ...       ...   ...
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0
46391  7360845   1  2.0  3.0  0.333333  0.666667   NaN
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0
46394  7360941   2  NaN  2.0  1.000000  0.000000   1.0

[46395 rows x 7 columns]

In [14]:
# How many times that the user make consumptions?
X['u7'] = X.u3 + X.u6.fillna(0)
X

User_id  u1   u2   u3        u4        u5    u6    u7
0          184   1  1.0  2.0  0.500000  0.500000   NaN   2.0
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0
2          687   2  6.0  8.0  0.250000  0.750000   NaN   8.0
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0
...        ...  ..  ...  ...       ...       ...   ...   ...
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0
46391  7360845   1  2.0  3.0  0.333333  0.666667   NaN   3.0
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0
46394  7360941   2  NaN  2.0  1.000000  0.000000   1.0   3.0

[46395 rows x 8 columns]

In [15]:
# What is the percentage that the customers use the coupon?
X['u8'] = X.u2 / X.u7
X.fillna(0.0, inplace=True)
X

User_id  u1   u2   u3        u4        u5    u6    u7        u8
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636
...        ...  ..  ...  ...       ...       ...   ...   ...       ...
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000

[46395 rows x 9 columns]

We also noticed that the gap of the user's consumption may help us to identify more features about the user, like how frequently that he/she will make a consumption, the ratio of coupon usage, the frequency in a defined period(**in 15 days**),etc. we will make a digging in such features.

In [16]:
# first of all, how frequently that a user to make a consumption in average?
temp = offline[offline.Date.notnull()]
temp['Date'] = pd.to_datetime(temp['Date'],format='%Y%m%d')
temp = pd.merge(temp, temp.groupby('User_id').Date.max().reset_index(name='_max'))
temp = pd.merge(temp, temp.groupby('User_id').Date.min().reset_index(name='_min'))
temp = pd.merge(temp, temp.groupby('User_id').size().reset_index(name='_len'))
temp['u9'] = (temp._max - temp._min).dt.days / temp._len
temp.drop_duplicates('User_id', inplace=True)
X = pd.merge(X, temp[['User_id', 'u9']],how='left',on='User_id')
X

C:\Users\ugz20\AppData\Local\Temp\ipykernel_18460\3781809539.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp['Date'] = pd.to_datetime(temp['Date'],format='%Y%m%d')


User_id  u1   u2   u3        u4        u5    u6    u7        u8  \
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000   
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429   
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000   
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857   
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636   
...        ...  ..  ...  ...       ...       ...   ...   ...       ...   
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667   
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222   
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000   

              u9  
0       0.000000  
1       7.333333  
2      31.500000  
3       5.600000  
4      19.428571  
...          ...  
46390  32.000000  
46391   0.000000  
46392   9.666667  
46393  17.428571  
46394   0.666667  

[46395 rows x 10 columns]

In [17]:
# How about their ordinary consumption gap?
temp = Ordinary_Consumption
temp['Date'] = pd.to_datetime(temp['Date'],format='%Y%m%d')
temp = pd.merge(temp, temp.groupby('User_id').Date.max().reset_index(name='_max'))
temp = pd.merge(temp, temp.groupby('User_id').Date.min().reset_index(name='_min'))
temp = pd.merge(temp, temp.groupby('User_id').size().reset_index(name='_len'))
temp['u10'] = (temp._max - temp._min).dt.days / temp._len
temp.drop_duplicates('User_id', inplace=True)
X = pd.merge(X, temp[['User_id', 'u10']], how='left', on='User_id')
X

C:\Users\ugz20\AppData\Local\Temp\ipykernel_18460\2482615655.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp['Date'] = pd.to_datetime(temp['Date'],format='%Y%m%d')


User_id  u1   u2   u3        u4        u5    u6    u7        u8  \
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000   
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429   
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000   
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857   
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636   
...        ...  ..  ...  ...       ...       ...   ...   ...       ...   
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667   
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222   
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000   

              u9        u10  
0       0.000000        NaN  
1       7.333333   4.000000  
2      31.500000        NaN  
3       5.600000   6.461538  
4      19.428571  27.000000  
...          ...        ...  
46390  32.000000  48.000000  
46391   0.000000        NaN  
46392   9.666667  14.500000  
46393  17.428571  19.500000  
46394   0.666667   0.000000  

[46395 rows x 11 columns]

In [18]:
# By the way, How about their coupon consumption gap??
coupon_consume['Date'] = pd.to_datetime(coupon_consume['Date'],format='%Y%m%d')
temp = pd.merge(coupon_consume, coupon_consume.groupby('User_id').Date.max().reset_index(name='_max'))
temp = pd.merge(temp, temp.groupby('User_id').Date.min().reset_index(name='_min'))
temp = pd.merge(temp, temp.groupby('User_id').size().reset_index(name='_len'))
temp['u11'] = (temp._max - temp._min).dt.days / temp._len
temp.drop_duplicates('User_id', inplace=True)
X = pd.merge(X, temp[['User_id', 'u11']], how='left', on='User_id')
X

C:\Users\ugz20\AppData\Local\Temp\ipykernel_18460\4040961787.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  coupon_consume['Date'] = pd.to_datetime(coupon_consume['Date'],format='%Y%m%d')


User_id  u1   u2   u3        u4        u5    u6    u7        u8  \
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000   
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429   
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000   
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857   
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636   
...        ...  ..  ...  ...       ...       ...   ...   ...       ...   
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667   
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222   
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000   

              u9        u10    u11  
0       0.000000        NaN   0.00  
1       7.333333   4.000000   0.00  
2      31.500000        NaN  31.50  
3       5.600000   6.461538  11.75  
4      19.428571  27.000000  52.50  
...          ...        ...    ...  
46390  32.000000  48.000000   0.00  
46391   0.000000        NaN   0.00  
46392   9.666667  14.500000   0.00  
46393  17.428571  19.500000   0.00  
46394   0.666667   0.000000   0.50  

[46395 rows x 12 columns]

Now is time to think about the coupon and features relationship, this can also bring some effective.

In [19]:
# How many full discount coupon that the customer receive?
temp = offline[offline.Discount_rate.str.contains(':') == True]
temp = temp.groupby('User_id').size().reset_index(name='u12')
X = pd.merge(X, temp, how='left', on='User_id')
X

User_id  u1   u2   u3        u4        u5    u6    u7        u8  \
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000   
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429   
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000   
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857   
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636   
...        ...  ..  ...  ...       ...       ...   ...   ...       ...   
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667   
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222   
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000   

              u9        u10    u11  u12  
0       0.000000        NaN   0.00  2.0  
1       7.333333   4.000000   0.00  5.0  
2      31.500000        NaN  31.50  8.0  
3       5.600000   6.461538  11.75  7.0  
4      19.428571  27.000000  52.50  6.0  
...          ...        ...    ...  ...  
46390  32.000000  48.000000   0.00  2.0  
46391   0.000000        NaN   0.00  3.0  
46392   9.666667  14.500000   0.00  3.0  
46393  17.428571  19.500000   0.00  3.0  
46394   0.666667   0.000000   0.50  2.0  

[46395 rows x 13 columns]

In [20]:
# How many discount coupon that the customer receive?
temp = offline[offline.Discount_rate.str.contains('\.') == True]
temp = temp.groupby('User_id').size().reset_index(name='u13')
X = pd.merge(X, temp, how='left', on='User_id')
X

User_id  u1   u2   u3        u4        u5    u6    u7        u8  \
0          184   1  1.0  2.0  0.500000  0.500000   0.0   2.0  0.500000   
1          417   1  4.0  5.0  0.200000  0.800000   2.0   7.0  0.571429   
2          687   2  6.0  8.0  0.250000  0.750000   0.0   8.0  0.750000   
3          696   4  5.0  9.0  0.444444  0.555556  26.0  35.0  0.142857   
4          947   2  4.0  6.0  0.333333  0.666667   5.0  11.0  0.363636   
...        ...  ..  ...  ...       ...       ...   ...   ...       ...   
46390  7360797   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46391  7360845   1  2.0  3.0  0.333333  0.666667   0.0   3.0  0.666667   
46392  7360863   1  2.0  3.0  0.333333  0.666667   2.0   5.0  0.400000   
46393  7360928   1  2.0  3.0  0.333333  0.666667   6.0   9.0  0.222222   
46394  7360941   2  0.0  2.0  1.000000  0.000000   1.0   3.0  0.000000   

              u9        u10    u11  u12  u13  
0       0.000000        NaN   0.00  2.0  NaN  
1       7.333333   4.000000   0.00  5.0  NaN  
2      31.500000        NaN  31.50  8.0  NaN  
3       5.600000   6.461538  11.75  7.0  2.0  
4      19.428571  27.000000  52.50  6.0  NaN  
...          ...        ...    ...  ...  ...  
46390  32.000000  48.000000   0.00  2.0  1.0  
46391   0.000000        NaN   0.00  3.0  NaN  
46392   9.666667  14.500000   0.00  3.0  NaN  
46393  17.428571  19.500000   0.00  3.0  NaN  
46394   0.666667   0.000000   0.50  2.0  NaN  

[46395 rows x 14 columns]

In [21]:
# make the output
X.to_csv("User_feature.csv")